### ISSUE TITLE
deployments (55 functions) failing with message "Build failed with status: EXPIRED..."

### ISSUE THREAD

**Issue by wojtek-iwarranty** [firebase/firebase-tools#7268]
### [REQUIRED] Environment info

**firebase-tools:**
13.8.0 or 13.10.2

**Platform:**
ubuntu-latest, via GitHubAction

### [REQUIRED] Test case

run `firebase deploy --force --only functions,firestore` with a codebase containing 55 2nd Cloud Functions.
In my case it is ran via action w9jds/firebase-action@v13.10.2 (or w9jds/firebase-action@v13.8.0)

### [REQUIRED] Steps to reproduce

Same as test case

### [REQUIRED] Expected behavior

I would expect this deployment to go through since it is below the deployment quota of 80 functions per minute.

### [REQUIRED] Actual behavior

Some of the functions fail to deploy with build with the following error:

Build failed with status: `EXPIRED and message: An unexpected error occurred. Refer to build logs: [redacted]. For more details see the logs at [redacted].`
The build logs are empty and don't contain any useful information other than the build expiring.

I can see in Cloud Build console that some of the builds failed:
![image](https://github.com/firebase/firebase-tools/assets/143395186/4b405b90-a18b-499f-85b8-323431a55249)

This happens for about 15 of the 55 functions. Here is a truncated output of the full deploy:

```
Run w9jds/firebase-action@v13.8.0
  with:
    args: deploy --force --config=firebase.json --only functions,firestore
  env:
    GCP_SA_KEY: ***
  
    PROJECT_ID: [redacted]
/usr/bin/docker run --name w9jdsfirebaseactionv1380_9e59c3 --label f3784f --workdir /github/workspace --rm -e  [redacted] -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/rabbit/rabbit":"/github/workspace" w9jds/firebase-action:v13.8.0 deploy --force --config=firebase.json --only functions,firestore
Storing GCP_SA_KEY in /opt/gcp_key.json
Exporting GOOGLE_APPLICATION_CREDENTIALS=/opt/gcp_key.json
setting firebase project to [redacted]
Now using alias [redacted] ([redacted])

=== Deploying to '[redacted]'...

i  deploying firestore, functions
i  firestore: reading indexes from apps/[redacted]/firestore.indexes.json...
i  cloud.firestore: checking apps/[redacted]/firestore.rules for compilation errors...
✔  cloud.firestore: rules file apps/[redacted]/firestore.rules compiled successfully
i  functions: preparing codebase [redacted]-functions for deployment
i  functions: ensuring required API cloudfunctions.googleapis.com is enabled...
i  functions: ensuring required API cloudbuild.googleapis.com is enabled...
i  artifactregistry: ensuring required API artifactregistry.googleapis.com is enabled...
✔  functions: required API cloudbuild.googleapis.com is enabled
✔  functions: required API cloudfunctions.googleapis.com is enabled
✔  artifactregistry: required API artifactregistry.googleapis.com is enabled
⚠  functions: package.json indicates an outdated version of firebase-functions. Please upgrade using npm install --save firebase-functions@latest in your functions directory.
⚠  functions: Please note that there will be breaking changes when you upgrade.
i  functions: Loading and analyzing source code for codebase [redacted]-functions to determine what to deploy
Serving at port 8640

Firebase Admin Version: 11.11.0

i  functions: preparing dist/apps/[redacted]-functions directory for uploading...
i  functions: packaged /github/workspace/dist/apps/[redacted]-functions (some MB) for uploading
i  functions: ensuring required API run.googleapis.com is enabled...
i  functions: ensuring required API eventarc.googleapis.com is enabled...
i  functions: ensuring required API pubsub.googleapis.com is enabled...
i  functions: ensuring required API storage.googleapis.com is enabled...
✔  functions: required API eventarc.googleapis.com is enabled
✔  functions: required API run.googleapis.com is enabled
✔  functions: required API pubsub.googleapis.com is enabled
✔  functions: required API storage.googleapis.com is enabled
i  functions: generating the service identity for pubsub.googleapis.com...
i  functions: generating the service identity for eventarc.googleapis.com...
i  firestore: deploying indexes...
i  firestore: latest version of apps/[redacted]/firestore.rules already up to date, skipping upload...
✔  firestore: deployed indexes in apps/[redacted]/firestore.indexes.json successfully for (default) database
✔  functions: dist/apps/[redacted]-functions folder uploaded successfully
✔  firestore: released rules apps/[redacted]/firestore.rules to cloud.firestore
i  functions: updating Node.js [2](https://github.com/iwarranty-io/rabbit/actions/runs/[redacted]#step:9:2)0 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success]full(europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
i  functions: updating Node.js 20 (2nd Gen) function [redacted]-functions:[redacted-success](europe-west1)...
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success]full(europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: https://console.cloud.google.com/cloud-build/builds;region=europe-west1/AA
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.
✔  functions[[redacted]-functions:[redacted-success](europe-west1)] Successful update operation.

Functions deploy had errors with the following functions:
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
	[redacted]-functions:[redacted](europe-west1)
i  functions: cleaning up build files...
⚠  functions: Unhandled error cleaning up build images. This could result in a small monthly bill if not corrected. You can attempt to delete these images by redeploying or you can delete them manually at https://console.cloud.google.com/artifacts/docker/[redacted]-[redacted]/europe-west1/gcf-artifacts
Error: There was an error deploying functions:
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
- Error Failed to update function [redacted] in region europe-west1
```

It's actually reopening for #5967 ...

**Comment by gersonlimadev** [comment]
The same errors are happening here to me

**Comment by toddmotto** [comment]
It amazes me how this is going to badly wrong again @joehan the previous 'fix' suggested by running `firebase experiments:enable functionsv2deployoptimizations` now does not work, what are we supposed to do to deploy vital function updates?

**Comment by raymond-tetteh** [comment]
Experiencing the same issues. Can't deploy my functions.

**Comment by onn-software** [comment]
Same here, 51 functions, and it worked for years. Now I have 'Build failed with status: EXPIRED and message: An unexpected error occurred'.

**Comment by halafi** [comment]
same issue, also sometimes deployment is skipped when there are changes to be deployed, thinking to --force deploy each function manually

**Comment by Endran** [comment]
We encounter the same. Breaking up the build in doesn't work either, each of the groups takes forever.

```
firebase deploy --only functions:group1
firebase deploy --only functions:group2
firebase deploy --only functions:group3
firebase deploy --except functions
```

Each function group contains around 10 functions, and takes over 10 minutes, where it used to deploy all in a few minutes. 

The last successful deploy was on May 30th, first failing deploy June 2nd. What did the Firebase/GCP team change in the meantime?

**Comment by wesleygonalv** [comment]
We are facing the same issue: we have over 100 functions, but we can only deploy a few of them.

We considered breaking them into smaller groups, but as @Endran mentioned, it unfortunately did not solve the problem.

We have been experiencing this issue since the date mentioned by others and are still waiting for updates.

**Comment by aalej** [comment]
Hey everyone, thanks for raising this and apologies for the issue this has caused. Our engineering team has been notified about this issue. 

While waiting for our engineering team’s response, I’m currently trying to replicate this by deploying ~60 v2 functions, but so far I haven’t been able to encounter the same error. I would like to ask if anyone could share the ff:

1. Do you have the experiment `functionsv2deployoptimizations` enabled?
2. Which region are you deploying the function to? E.g. `europe-west1`, `us-central1`, etc.
3. What functions are being deployed?  E.g. `onRequest`, `beforeUserSignedIn`, `onDocumentWritten`, etc.

**Comment by wesleygonalv** [comment]
Hi @aalej, about your questions.

Do you have the experiment functionsv2deployoptimizations enabled?
> As far as I know, we're not using it. Is there a command to check?
Which region are you deploying the function to? 
> europe-west3
What functions are being deployed? E.g. onRequest, beforeUserSignedIn, onDocumentWritten, etc.
> We have `onCall`,  `onRequest` and `onMessagePublished`, all functions for v2

**Comment by charifield** [comment]
@aalej Thanks for looking into this:
1. Yes, it is enabled
2. us-central1
3. 100 v2 HTTP `onRequest` functions, and about 3 `onValueWritten` functions for the RTD

**Comment by aalej** [comment]
Thanks for the additional information folks! @wesleygonalv, you can check if `functionsv2deployoptimizations` by running:
```
firebase experiments:describe functionsv2deployoptimizations
```

**Comment by wesleygonalv** [comment]
@aalej Thanks for the support,

We are not using functions v2 deploy optimizations. Is there anything else I can help you with?

**Comment by wojtek-iwarranty** [comment]
Hi @aalej  thx for your comment, it gives us hope 😉

> 1. Do you have the experiment `functionsv2deployoptimizations` enabled?
```
Name: functionsv2deployoptimizations
Enabled: yes
```
> 2. Which region are you deploying the function to? E.g. `europe-west1`, `us-central1`, etc.

`europe-west1` and `australia-southeast1`
> 3. What functions are being deployed?  E.g. `onRequest`, `beforeUserSignedIn`, `onDocumentWritten`, etc.

45 x `onCall` (http) + 10 x `onDocumentWritten` (firestore)

**Comment by wojtek-iwarranty** [comment]
Hi @aalej , maybe this will help you somehow. After last weekend the new "success" number for `Per project mutation requests per minute per region` is around 40 (when before it was 55 so all our functions).
![image](https://github.com/firebase/firebase-tools/assets/143395186/e3b6f57d-8219-4613-8a53-e433238204e9)

**Comment by tbusot** [comment]
Region: `us-central-1`

Expiremental Optimizations:
```
Name: functionsv2deployoptimizations
Enabled: yes
Description: Reuse build images across funtions to increase performance and reliaibility of deploys. This has been made an experiment due to backend bugs that are temporarily causing failures in some regions with this optimization enabled
```

Function Breakdown: 

35 `onDocumentCreated`, `onDocumentUpdated`, and `onDocumentWritten`
2 `onSchedule` (Adding these is when I started seeing the issues, whether it was causal or coincidence is unknown as now even when I remove them, I still get the errors.)
1 `onCreate` Auth
4 `onCall`
1 `Expo` set of API calls.

**Comment by Endran** [comment]
> Do you have the experiment functionsv2deployoptimizations enabled?
No, trying now, will update this asap

> Which region are you deploying the function to? E.g. europe-west1, us-central1, etc.
`europe-west1`

> What functions are being deployed? E.g. onRequest, beforeUserSignedIn, onDocumentWritten, etc.

11x onMessagePublished
19x onCall
3x onRequest
1x v1 auth onCreate
1x v1 auth onDelete
13x onDocumentWritten
4x onDocumentCreated
1x onDocumentUpdated
1x onDocumentDeleted
5x onSchedule

**Comment by aalej** [comment]
Hey folks, just a quick update here. I’m able to reproduce the issue. Thanks for all the information you’ve provided! I’ll mark this issue as reproducible.

**Comment by Endran** [comment]
Any update on this? It's blocking our releases for a week now, which impacts business. Some relevant logs from the build.

```
...
success Installed "firebase-tools@13.11.1" with binaries:
      - firebase
Done in 22.28s.
$ firebase experiments:enable functionsv2deployoptimizations
Enabled experiment 
functionsv2deployoptimizations
$ firebase experiments:describe functionsv2deployoptimizations
Name: functionsv2deployoptimizations
Enabled: yes
...
49x functions: updating Node.js 18 (2nd Gen) function ...
2x functions: updating Node.js 18 (1st Gen) function auth-...
...
12x: Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs: ...
...
Error: There was an error deploying functions:
12x: Error: Failed to update function
```

Please keep us posted on progress on this blocking issue.

**Comment by wesleygonalv** [comment]
@SherpaMiguel

We are going through the same thing, and one of the temporary solutions we found was to deploy 4 or 5 functions at a time, we had problems in the production environment and the solution to upload the fixes was this.

```bash
firebase deploy --only "functions:functionName, ..."
```

This command can be run on other cloud build runs at the same time 👍

**Comment by omgovich** [comment]
Having the same issue (`Build failed with status: EXPIRED and message`).
We can't deploy anything (it always fails) and just loosing money because of this bug.
`functionsv2deployoptimizations` doesn't help.

We deploy 38 functions to `us-central1`. Most of them are v2 `onCall`.

The interesting thing is that our another project, which has 85 functions, has no issues with deployment. Most of the functions there are v1. So I assume the issue is related to v2.

**Comment by curtis-jotson** [comment]
I'm going to just +1 this. We've been fighting this for weeks. It started in our CD pipeline and now is doing it even with our manual deploys. This might already be known or figured out, but I'll share what I know so far. I am by no means a GCP/Firebase expert or team member so take this for what it is.

Firebase deploy uses Cloud Build to actually build the function containers that run on GCP. These builds have a default (I think) `queueTtl` configuration. This configuration controls how long a function build sits in the build queue before the build itself expires. It's either configured for or defaulting to 5/6 minutes. Per the Cloud Build documentation, the way I understand it, the default worker pool that actually builds the functions should be limited to 30 concurrent builds (https://cloud.google.com/build/docs/private-pools/private-pools-overview). What we're seeing in the GCP console is that it's only running 5 at a time. We have ~30 functions that are taking ~1-2 minutes per build. At a concurrency of 5, the ones batched last are expiring because they're sitting in the queue for longer than the 5/6 minutes that `queueTtl` is set to.

In addition to all this, when we do get a failing deploy, subsequent deploys using the functionality that only deploys the functions that need to be updated says all functions are up to date. However upon checking functionally, they are not and have not been deployed.

So my thinking on possibilities of things that have changed are:
- GCP changed concurrent builds on the default worker pool from 30 to 5
- Something about the Firebase Function builds is taking longer than it used to
- The `queueTtl` value has changed
- The default build runner has been nerfed?
- It used to be doing 1 build and now it's doing 1 for each function?
- Just throwing it out there but it's possible we've added an external dependency that is extending the build time drastically

As far as I can tell (with my limited Google-Fu), with `firebase deploy` we are unable to configure the `queueTtl` value or the runner type that the builds run on.

**Comment by cedvdb** [comment]
> 
> 1. Do you have the experiment `functionsv2deployoptimizations` enabled?
> 2. Which region are you deploying the function to? E.g. `europe-west1`, `us-central1`, etc.
> 3. What functions are being deployed?  E.g. `onRequest`, `beforeUserSignedIn`, `onDocumentWritten`, etc.

1. Yes
2. europe-west-1
3. Firestore triggers, Scheduled functions, Storage Triggers, and an on request trigger


It is successful when forcing a second run

**Comment by AvishJ03** [comment]
Facing same issue since a week. Any fixes yet? Tried with both `functionsv2deployoptimizations` enabled and disabled, not working for both. The region is us-central1 and we are usng onRequest for v2 functions

**Comment by toddmotto** [comment]
@SherpaMiguel

Yeah, this is really poor. Considering a fix was previously published and worked absolutely fine, and now broken again with zero fix you have to question what they are even doing.

**Comment by Endran** [comment]
I tried to contact Firebase Support at https://firebase.google.com/support/troubleshooter/contact?authuser=0 . But they also just mention the same dead end about functionsv2deployoptimizations , and go quiet after the first response. Please try to contact them as well, in hopes the priority is raised.

**Comment by al4nnw** [comment]
Adding a +1 here.

So far, I've seen this issue in the past and searching google I got back to #5967, but this time going back to 13.8.0 didn't fixed it. 

Also, as @onn-software said, the build times went from success in 6 minutes to a failure at 12, which could be related to these expiration errors.

**Comment by mkdior** [comment]
> * Just throwing it out there but it's possible we've added an external dependency that is extending the build time drastically

I really doubt that, the occurrences of this issue exploded during a specific week, hence the activity in this issue, I think something changed internally that started to cause this. I doubt this many people introduced something affecting their build times drastically in the same period of time.

Wish we could introduce project wide defaults for GCP, allowing us to set timeouts for the cloud builder for example. Anyways, in essence, +1, sick and tired of this issue.

**Comment by hasbean** [comment]
+1.  I'm not sure how Firebase team doesn't see this as a top priority.  I've been stuck with undeployed functions for weeks now.

**Comment by toddmotto** [comment]
I'm baffled too that this isn't 1) a high priority and 2) already fixed within a day or so. This is a fundamental building block of Firebase and with no update in days and days it's difficult to question the reliability of the platform. 

It's also not like this is a new 'bug' that needs tracking down, refactoring and deploying, they literally already fixed it - and broke it again.

**Comment by cedvdb** [comment]
@toddmotto it's probably high priority, but we don't know for sure since there is no label on this issue. 

The only thing you can do is +1 the top comment and contact [their support](https://firebase.google.com/support/troubleshooter/contact?authuser=0 ) to increase the priority.

**Comment by mandalornl** [comment]
> Hi everyone,
> 
> We are facing the same problem with deploying our Cloud Functions.
> 
> Is there any update or known workaround for this issue? Any help would be greatly appreciated.
> 
> Thank you!

As a workaround you can try deploying in small batches. This has worked for me. I'd suggest 5 functions at a time. The only caveat is, it'll take forever to complete when you have many many many functions to deploy :sleeping: But you can easily automate the process if you know you way around e.g. Node.js.

**Comment by CarlosDeucherMoretoCont** [comment]
+1 I am facing the same issue here, implementing our 88 Cloud Functions ):

**Comment by gersonlimadev** [comment]
I'm facing the same issue for almost 2 weeks now, it's crazy
<img width="279" alt="image" src="https://github.com/firebase/firebase-tools/assets/485899/af537178-f7b4-427a-a5c4-40b5d8a2d6f4">

**Comment by blidd-google** [comment]
Hey folks, really sorry to hear about all the build-related errors you're seeing. We've been investigating the issue and reached out to our Cloud partner teams to understand why builds are expiring at an abnormally high rate.

We recently had to disable the single builds feature, which allowed multiple functions to re-use builds and thus significantly reduces the number of builds per function deployment. I suspect that the rollback has compounded the severity of the expiration problem you all are facing, and we're actively working on re-enabling the feature.

Again, sorry to hear about these deployment issues; we appreciate your patience, and I'll keep this thread updated on our progress, both for our investigation into the underlying issue causing the number of concurrent builds / queue TTLs to decrease significantly and for our effort to safely re-enable single builds.

**Comment by kdawgwilk** [comment]
Are there any workarounds we can implement ourselves other than breaking the function deploys into groups of 5?  Is there a way for us to increase the timeouts? If we built a single function and deployed it would the next deploy be able to use the first function build cache?

**Comment by toddmotto** [comment]
2 weeks ago this issue was opened... unable to deploy still, lost for words a bit on this.

**Comment by cedvdb** [comment]
Deploying usually works when running the deployment multiple times back to back. So you can make a shell script that runs firebase deploy 3 times. It will take longer but you'll be able to deploy

**Comment by mkdior** [comment]
> Deploying usually works when running the deployment multiple times back to back. So you can make a shell script that runs firebase deploy 3 times. It will take longer but you'll be able to deploy

Each deployment consumes builder minutes as far as I'm aware, which costs us. There should be no extra costs incurred due to a bug.

**Comment by wesleygonalv** [comment]
Now some of my attempts are returning this, will we have an update for the expired problem?

`
Error: Cloud Runtime Config is currently experiencing issues, which is preventing your functions from being deployed. Please wait a few minutes and then try to deploy your functions again."
`

**Comment by GabiGrin** [comment]
Same here

**Comment by Endran** [comment]
Ahhh.. I have a high prio issue I need to push, and instead of 10 minutes build and 5 minutes deploy (front end and backend) it takes 10 + 20 minutes to go live. I broke up everything in smaller batches, but it's taking ages. Its been 3 weeks now since our last normal deploy, please fix it asap.

**Comment by wesleygonalv** [comment]
We have a similar case @Endran, 

we had to deploy more than once to correct production problems, with the risk of losing the partner company :(

**Comment by MNITD** [comment]
Hey @blidd-google

It's a week since your last message. Is there any bad or good news regarding the resolution of the issue?

Also, I believe many people will be happy to hear an estimate on when we can expect at least a workaround (as it was with downgrading the CLI to 13.8.0)

**Comment by tbusot** [comment]
For those stuck with this for, you can deploy in 2 steps reliably when you do a manual deploy, rather than running the job over and over, hoping it succeeds eventually:

First, do a `firebase deploy --only functions` 

If it fails (and it probably will) it's going to fail on some, succeed on *hopefully* most.
At the end of the run, it's going to produce a list of functions that failed to deploy like this:

```
- Error Failed to update function bioCreated in region us-central1
- Error Failed to update function messageCreated in region us-central1
- Error Failed to update function groupInvitationUpdated in region us-central1
```

Extract that list of functions into a list and re-run deploy like this:
`firebase deploy --only "functions:bioCreated, functions:messageCreated, functions:groupInvitationUpdated"` and it should succeed.

**Comment by blidd-google** [comment]
Hey everyone, we're still waiting for the single builds feature to finish rolling out globally - once we do, we'll update the CLI to re-enable the feature by default. The Cloud team has been proceeding carefully with the release given that we've had issues with the feature in the past. A rough estimate would be at least another week until the global rollout is complete.

> Are there any workarounds we can implement ourselves other than breaking the function deploys into groups of 5? Is there a way for us to increase the timeouts? 

Unfortunately, users are unable to modify the Cloud Build queue TTL. We're considering increasing the TTL for all GCF 2nd gen functions, and will update if we decide to move forward with that. 

> If we built a single function and deployed it would the next deploy be able to use the first function build cache?

You are essentially describing the Single Builds feature. :)

> Each deployment consumes builder minutes as far as I'm aware, which costs us. There should be no extra costs incurred due to a bug.

To clarify, [build-minutes are not incurred for the time that a build is in the queue](https://cloud.google.com/build/pricing).

> Now some of my attempts are returning this, will we have an update for the expired problem? `Error: Cloud Runtime Config is currently experiencing issues, which is preventing your functions from being deployed. Please wait a few minutes and then try to deploy your functions again."`

This is a separate issue and has been mitigated. https://github.com/firebase/firebase-tools/issues/7341 

You can also [request a quota increase](https://cloud.google.com/build/quotas#increasing_quota) for the number of concurrent builds via the Google Cloud Console. Truly sorry for the continued deployment issues. I'm confident that the single builds feature will mitigate the build expiration issue, but it has caused issues in the past and we'd like to make sure we do it right this time.

**Comment by megastep** [comment]
Still failing randomly with expired builds for me here... I've had a few instances of builds successfully completing but it's far from being all fixed as far as I can tell, at least as far as my projects go.

I've also requested and been granted quota increases and it doesn't really seem to make the problem go away.

**Comment by mkdior** [comment]
> Still failing randomly with expired builds for me here... I've had a few instances of builds successfully completing but it's far from being all fixed as far as I can tell, at least as far as my projects go.
> 
> I've also requested and been granted quota increases and it doesn't really seem to make the problem go away.

Same, I think the people who think it's fixed are re-running cached builds or something...

**Comment by wesleygonalv** [comment]
I am using the latest version of Firebase Tools, and in my case, the problem still persists.

![image](https://github.com/firebase/firebase-tools/assets/19338037/83a12423-3283-4799-a865-c70f3efb768c)

If I try to deploy again, the cache activates and indicates that the functions are updated. However, when I inspect the source code of the functions, the previous version with the old bugs is still there.

![image](https://github.com/firebase/firebase-tools/assets/19338037/cfa557e4-bd60-4e3d-8e36-c9742fdc1deb)

We urgently need to update a large set of functions, and it is proving to be very challenging to deploy them bit by bit.

**Comment by blidd-google** [comment]
Hey folks, some of you may have already experienced improved build performance due to the single builds feature being re-enabled. The feature is almost fully rolled out, but some of you may continue to encounter the build expiration issue until after the holidays.

Note that for firebase-tools >= 13.8.1, you must enable the feature explicitly by running `firebase experiments:enable functionsv2deployoptimizations`. Again, single builds hasn't completed its global rollout so some of you may still fail to build and deploy all your functions; but once the rollout is complete (likely at the beginning of next week), we will cut a new release of firebase-tools that will enable the feature again by default.

**Comment by matt-oakes** [comment]
Thank you for the update. To be clear, are you recommending that developers using the latest CLI should enable the experiment to have the best chance of resolving this issue?

I currently have it disable and usually a single function fails to deploy each time. Would this be resolved by enabling the experiment?

**Comment by mkdior** [comment]
> Thank you for the update. To be clear, are you recommending that developers using the latest CLI should enable the experiment to have the best chance of resolving this issue?
> 
> I currently have it disable and usually a single function fails to deploy each time. Would this be resolved by enabling the experiment?

Go ahead and try it, for me it seems to work, just ran a deployment set of more than 50 functions, all deployed, no timeouts. Some seem to be in invalid state, still, but at least I can get 40~ OK deployments this way.

**Comment by matt-oakes** [comment]
> Go ahead and try it, for me it seems to work, just ran a deployment set of more than 50 functions, all deployed, no timeouts. Some seem to be in invalid state, still, but at least I can get 40~ OK deployments this way.

What do you mean by invalid state? I'm a bit loathed to turn this on if you're saying 20% of your functions do not deploy properly or do not work after deploying.

With the experiement disabled I'm getting 1/50 failing to deploy, and it's usually a different one each time.

**Comment by mkdior** [comment]
> > Go ahead and try it, for me it seems to work, just ran a deployment set of more than 50 functions, all deployed, no timeouts. Some seem to be in invalid state, still, but at least I can get 40~ OK deployments this way.
> 
> What do you mean by invalid state? I'm a bit loathed to turn this on if you're saying 20% of your functions do not deploy properly or do not work after deploying.
> 
> With the experiement disabled I'm getting 1/50 failing to deploy, and it's usually a different one each time.

Invalid state as in, the CLI reports that the function has been "deployed", but google cloud states that the deployment failed.

**Comment by tbusot** [comment]
FYI - Upgraded to CLI version 13.13.0
Deployed 47 functions in one step with no errors.
I'm using the `firebase experiments:enable functionsv2deployoptimizations` flag.

**Comment by blidd-google** [comment]
Single builds has now been fully rolled out - to enable the feature in the CLI, run `firebase experiments:enable functionsv2deployoptimizations`. We'll be cutting a release next week to enable the feature again by default. Thanks for your patience, I'll continue to monitor this thread for problems but I expect that the vast majority of you will see significant improvements in your deploy performance.

**Comment by charlesfries** [comment]
Definitely working better with that experiment enabled, but I'm still seeing the expired message for 21 of my 98 V2 functions. Using `firebase-tools@13.13.0`

Is 98 simply too many for GCP to handle at once?

**Comment by andrewkrippner** [comment]
Still having problems. Experiment turned on, just had 8 out of 22 fail.

**Comment by chriswaterbury** [comment]
I'm still experiencing this with functionsv2deployoptimizations enabled. 8/51 functions "Build failed with status: EXPIRED"

**Comment by toddmotto** [comment]
Some feedback:

I have **not** upgraded the Firebase Tools version, I'm running `v13.10.0` and I just deployed - having **previously enabled** the `functionsv2deployoptimizations` patch weeks ago when this issue first arrived - and successfully deployed all my functions which were previously failing:
 
![Screenshot 2024-07-05 at 00 54 20](https://github.com/firebase/firebase-tools/assets/1655968/a70bedce-dc8c-4183-b6a6-679433256960)

**Comment by halafi** [comment]
why is this closed? I can still reproduce the issue

**Comment by mkdior** [comment]
And you’re sure the experimental deployment feat is enabled? What’s your
firebase tools version at?

**Comment by wesleygonalv** [comment]
Why is this closed? 

I still continue to have the problem even with the latest version and the feature enabled.

Manual deployments keep returning quota problems, 
![image](https://github.com/user-attachments/assets/0e9c4a84-c1d7-4f13-a62c-ec5983067be2)
while in the cloud build they return expired
![image](https://github.com/user-attachments/assets/1e526d39-b724-40f0-9e8e-1f3a9b3110ce)
This is a long list of errors
![image](https://github.com/user-attachments/assets/589e172c-a613-40f4-9347-a0f9362997dc)

Details:
firebase-tools - 13.13.2
Google Cloud SDK 478.0.0
bq 2.1.5
core 2024.05.24
gcloud-crc32c 1.0.0
gsutil 5.29
Node v18.17.0
Request for quota increase already requested and applied ✅
The `firebase experiments:enable` is enablead

**Comment by toddmotto** [comment]
It was a bit strange how this issue was marked as fixed and instantly closed, without gaining any feedback from users on here experiencing the issues to confirm the patch did actually fix the problem.

**Comment by chriswaterbury** [comment]
Still experiencing this intermittently with 5 or so functions failing out of ~60 functions

**Comment by Nushio** [comment]
This just happened to us right now, so it's very much still an issue.

> Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs:

**Comment by hasbean** [comment]
This is still happening from time to time.

**Comment by apiverveHQ** [comment]
Aaaaaaaand it's started again

`Build failed with status: EXPIRED and message: An unexpected error occurred. Refer to build logs`

**Comment by Mnumzane** [comment]
We are also still experiencing this issue with a set of v1 and v2 functions.

Will the build optimization from `firebase experiments:enable functionsv2deployoptimizations` not work when we have both v1 and v2 functions being deployed simultaneously?

**Comment by roddev-v** [comment]
Hey everybody. I'm joining in with my current situation.

I'was trying to deploy 2 HTTP V2 functions using firebase-tools@13.15.1 and I encounter mostly two outcomes:
- Build failed with status: FAILURE and message: An unexpected error occurred
- A "successful" build where my functions are skipped and in the Firebase Management console are marked as failed deployment.

I wasn't able to deploy a simple REST function with a clean set-up on macOS 14.5 / Node 18.20.4 or 20.9.0.
`firebase experiments:enable functionsv2deployoptimizations` does not make any difference in my case.
The failed Cloud Build build job fails without any visible logs.

I tried to deploy in both `europe-west3` and `us-central1` regions.

**Comment by taeold** [comment]
Firebase here. We don't get alerts on closed issue. The original issue was closed because the the specific platform issue that may have caused the original author's deployment to fail has been addressed.

I highly encourage folks to open up a new issue following the template. If you think there are specific issues with your project, feel free to open up a case at [Firebase Support](https://firebase.google.com/support). This will give us more visibility into your specific project.

I'm currently writing up a response in https://github.com/firebase/firebase-tools/issues/7514 - do keep an eye on this if you think you have a issue related to this one.
